utils/cmd/cmd.c

```diff
@@ -0,0 +1,721 @@
+	{ "softvol", softvol, "PCM_PATH y|n" },
+	{ 0 }
+};
+
+static const struct option long_options[] = {
+    {"dbus",        required_argument, NULL, 'B'},
+    {"version",     no_argument, NULL, 'V'},
+    {"help",        no_argument, NULL, 'h'},
+	{"quiet",       no_argument, NULL, 'q'},
+    { 0 },
+};
+
+static struct {
+	const char *progname;
+	bool quiet;
+	char service[32];
+} options = {
+	.progname = NULL,
+	.quiet = false,
+	.service = BLUEALSA_SERVICE,
+};
+
+static struct ba_dbus_ctx dbus_ctx;
+
+static void print_error(const char *format, ...) {
+	if (!options.quiet) {
+		va_list ap;
+		va_start(ap, format);
+		vfprintf(stderr, format, ap);
+		fputs("\n", stderr);
+		va_end(ap);
+	}
+}
+
+static void print_error_usage(const char *format, ...) {
+	if (!options.quiet) {
+		va_list ap;
+		va_start(ap, format);
+		vfprintf(stderr, format, ap);
+		va_end(ap);
+		fprintf(stderr, "\nTry '%s --help' for more information.\n", options.progname);
+	}
+}
+
+static bool check_path(const char *path) {
+	struct ba_pcm *pcms = NULL;
+	size_t pcms_count = 0;
+	DBusError err = DBUS_ERROR_INIT;
+	bool result = false;
+
+	if (!bluealsa_dbus_get_pcms(&dbus_ctx, &pcms, &pcms_count, &err)) {
+		print_error("Couldn't get BlueALSA PCM list: %s", err.message);
+		return false;
+	}
+
+	size_t i;
+	for (i = 0; i < pcms_count; i++)
+		if (strcmp(pcms[i].pcm_path, path) == 0) {
+			result = true;
+			break;
+		}
+
+	if (!result)
+		print_error("Invalid pcm path: %s", path);
+
+	free(pcms);
+	return result;
+}
+
+static bool get_pcm(const char *path, struct ba_pcm *pcm) {
+	struct ba_pcm *pcms = NULL;
+	bool found = false;
+	size_t pcms_count = 0;
+	size_t i;
+	DBusError err = DBUS_ERROR_INIT;
+
+	if (!bluealsa_dbus_get_pcms(&dbus_ctx, &pcms, &pcms_count, &err))
+		return false;
+
+	for (i = 0; i < pcms_count; i++)
+		if (strcmp(pcms[i].pcm_path, path) == 0) {
+			memcpy(pcm, &pcms[i], sizeof(*pcm));
+			found = true;
+			break;
+		}
+	free(pcms);
+	return found;
+}
+
+static int list_pcms(int argc, char *argv[]) {
+	if (argc > 1) {
+		print_error_usage("Too many arguments.");
+		return EXIT_FAILURE;
+	}
+
+	struct ba_pcm *pcms = NULL;
+	size_t pcms_count = 0;
+	DBusError err = DBUS_ERROR_INIT;
+
+	if (!bluealsa_dbus_get_pcms(&dbus_ctx, &pcms, &pcms_count, &err)) {
+		print_error("Couldn't get BlueALSA PCM list: %s", err.message);
+		return EXIT_FAILURE;
+	}
+
+	size_t i;
+	for (i = 0; i < pcms_count; i++)
+		printf("%s\n", pcms[i].pcm_path);
+
+	free (pcms);
+
+	return EXIT_SUCCESS;
+}
+
+static int get_codecs(int argc, char *argv[]) {
+	if (argc != 2) {
+		print_error_usage("Too many arguments.");
+		return EXIT_FAILURE;
+	}
+
+	const char *path = argv[1];
+
+	if (!check_path(path)) {
+		return EXIT_FAILURE;
+	}
+
+	DBusMessage *msg;
+	DBusMessage *rep;
+	DBusError err = DBUS_ERROR_INIT;
+	if ((msg = dbus_message_new_method_call(dbus_ctx.ba_service, path,
+					BLUEALSA_INTERFACE_PCM, "GetCodecs")) == NULL) {
+		print_error("Out of memory");
```

Comment:
For out of memory I'm often using `strerror(ENOMEM)` so the error text will be consistent among applications.

Reply:
see commit 4b97de6